src/profile-logic/transforms.js

```diff
@@ -399,6 +445,143 @@ export function mergeFunction(
   });
 }
 
+export function collapseLibrary(
+  thread: Thread,
+  resourceIndexToCollapse: IndexIntoResourceTable
+): Thread {
+  const { stackTable, funcTable, frameTable, resourceTable, samples } = thread;
+  const resourceNameIndex = resourceTable.name[resourceIndexToCollapse];
+  const newFrameTable: FrameTable = {
+    address: frameTable.address.slice(),
+    category: frameTable.category.slice(),
+    func: frameTable.func.slice(),
+    implementation: frameTable.implementation.slice(),
+    line: frameTable.line.slice(),
+    optimizations: frameTable.optimizations.slice(),
+    length: frameTable.length,
+  };
+  const newFuncTable: FuncTable = {
+    address: funcTable.address.slice(),
+    isJS: funcTable.isJS.slice(),
+    name: funcTable.name.slice(),
+    resource: funcTable.resource.slice(),
+    fileName: funcTable.fileName.slice(),
+    lineNumber: funcTable.lineNumber.slice(),
+    length: funcTable.length,
+  };
+  const newStackTable: StackTable = {
+    length: 0,
+    prefix: [],
+    frame: [],
+  };
+  const oldStackToNewStack: Map<
+    IndexIntoStackTable | null,
+    IndexIntoStackTable | null
+  > = new Map();
+  const prefixStackToCollapsedStack: Map<
+    IndexIntoStackTable | null, // prefix stack index
+    IndexIntoStackTable | null // collapsed stack index
+  > = new Map();
+  const stackToCollapsedFrame: Map<
+    IndexIntoStackTable | null,
+    IndexIntoFrameTable | null
+  > = new Map();
+
+  oldStackToNewStack.set(null, null);
+  prefixStackToCollapsedStack.set(null, null);
+  stackToCollapsedFrame.set(null, null);
+
+  for (let stackIndex = 0; stackIndex < stackTable.length; stackIndex++) {
+    const prefix = stackTable.prefix[stackIndex];
+    const frameIndex = stackTable.frame[stackIndex];
+    const funcIndex = frameTable.func[frameIndex];
+    const resourceIndex = funcTable.resource[funcIndex];
+    const newStackPrefix = oldStackToNewStack.get(prefix);
+
+    if (newStackPrefix === undefined) {
+      throw new Error('newStackPrefix must not be undefined');
+    }
+    if (resourceIndex === resourceIndexToCollapse) {
+      // The stack matches this resource.
+      const newPrefixFrame = stackToCollapsedFrame.get(newStackPrefix);
+      if (newPrefixFrame === undefined) {
+        // The prefix is not a collapsed stack. Now check for an existing collapsed
+        // stack at this level. If it exists we can use that.
```

Comment:
How about:
```
// The prefix is not a collapsed stack. So this stack will not collapse into its prefix stack.
// But it might collapse into a sibling stack, if there exists a sibling with the same resource.
// Check if a collapsed stack with the same prefix (i.e. a collapsed sibling) exists.
```